File: packages/sdk-py/src/agenttool/vault.py
"""Vault client for the agent-vault API (agent-vault.fly.dev).

Two encryption paths (per migration 0022_vault_agent_encrypted.sql):

  Default (server-encrypted at rest):
    .put(name, value, ...) — server encrypts; in-process runtime can read.

  Opt-in (zero-knowledge):
    .put_encrypted(name, plaintext, k_vault=, ...) — SDK encrypts before
    send; agenttool stores ciphertext only.
    .get_decrypted(name, k_vault=, ...) — fetches and decrypts locally
    (transparently falls through to plaintext if the secret was stored
    via the default path).

Use put_encrypted for secrets agenttool itself shouldn't be able to read
(personal data, sensitive credentials you don't want the runtime to use).
Use put for secrets the hosted runtime needs to consume server-side
(e.g. LLM provider API keys).
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

import httpx

from .crypto import decrypt_thought, encrypt_thought
from .exceptions import AgentToolError


class VaultClient:
    """Client for the agent-vault API — AES-256-GCM encrypted secrets.

    Usage::

        at = AgentTool()

        # Store a secret
        at.vault.put("openai-key", "sk-...")

        # Retrieve it
        secret = at.vault.get("openai-key")
        print(secret["value"])

        # List names (values never returned in list)
        names = at.vault.list()

        # Delete
        at.vault.delete("openai-key")

        # Version history
        versions = at.vault.versions("openai-key")

        # Audit log
        events = at.vault.audit("openai-key")

        # Set access policy
        at.vault.set_policy("openai-key", allowed_agents=["agent-1"], read_only=True)

        # Bulk retrieve
        secrets = at.vault.bulk(["openai-key", "db-url"])

        # Check existence
        exists = at.vault.check(["openai-key", "missing-key"])
    """

    def __init__(self, http: httpx.Client, base_url: str) -> None:
        self._http = http
        self._base = base_url.rstrip("/")

    def _url(self, path: str) -> str:
        return f"{self._base}{path}"

    # ── Core CRUD ─────────────────────────────────────────────────────────────

    def put(
        self,
        name: str,
        value: str,
        *,
        description: Optional[str] = None,
        agent_ids: Optional[List[str]] = None,
        tags: Optional[List[str]] = None,
        ttl_seconds: Optional[int] = None,
        rotation_days: Optional[int] = None,
        agent_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Store or update a secret (AES-256-GCM encrypted at rest).

        Args:
            name: Secret name (slug-style, e.g. ``openai-key``).
            value: Plaintext secret value.
            description: Optional human-readable description.
            agent_ids: List of agent IDs allowed to access this secret.
            tags: Optional tags for organisation.
            ttl_seconds: Auto-expire after this many seconds.
            rotation_days: Flag for rotation after N days.
            agent_id: Pass ``X-Agent-Id`` header for audit trail.
        """
        payload: Dict[str, Any] = {"value": value}
        if description is not None:
            payload["description"] = description
        if agent_ids is not None:
            payload["agent_ids"] = agent_ids
        if tags is not None:
            payload["tags"] = tags
        if ttl_seconds is not None:
            payload["ttl_seconds"] = ttl_seconds
        if rotation_days is not None:
            payload["rotation_days"] = rotation_days

        headers = {}
        if agent_id is not None:
            headers["X-Agent-Id"] = agent_id

        resp = self._http.put(
            self._url(f"/v1/vault/{name}"), json=payload,
            headers=headers if headers else None,
        )
        if resp.status_code not in (200, 201):
            raise AgentToolError(f"vault.put failed: {resp.status_code}", hint=resp.text)
        return resp.json()

    def get(
        self,
        name: str,
        *,
        version: Optional[int] = None,
        agent_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Retrieve a secret's plaintext value.

        Args:
            name: Secret name.
            version: Retrieve a specific version (default: latest).
            agent_id: Pass ``X-Agent-Id`` header for audit trail.

        Returns dict with ``name``, ``value``, ``version``, ``description``, etc.
        """
        params: Dict[str, Any] = {}
        if version is not None:
            params["version"] = version

        headers = {}
        if agent_id is not None:
            headers["X-Agent-Id"] = agent_id

        resp = self._http.get(
            self._url(f"/v1/vault/{name}"),
            params=params or None,
            headers=headers if headers else None,
        )
        if resp.status_code == 404:
            raise AgentToolError("secret not found", hint=f"name={name}")
        if resp.status_code != 200:
            raise AgentToolError(f"vault.get failed: {resp.status_code}", hint=resp.text)
        return resp.json()

    def delete(self, name: str) -> Dict[str, Any]:
        """Soft-delete a secret (all versions)."""
        resp = self._http.delete(self._url(f"/v1/vault/{name}"))
        if resp.status_code != 200:
            raise AgentToolError(f"vault.delete failed: {resp.status_code}", hint=resp.text)
        return resp.json()

    def list(
        self,
        *,
        tag: Optional[str] = None,
        expiring_soon: Optional[bool] = None,
        rotation_due: Optional[bool] = None,
    ) -> List[Dict[str, Any]]:
        """List all secrets (names and metadata — values are never returned).

        Args:
            tag: Filter by tag.
            expiring_soon: If True, only return secrets expiring within 7 days.
            rotation_due: If True, only return secrets past their rotation date.
        """
        params: Dict[str, Any] = {}
        if tag is not None:
            params["tag"] = tag
        if expiring_soon is not None:
            params["expiring_soon"] = "true" if expiring_soon else "false"
        if rotation_due is not None:
            params["rotation_due"] = "true" if rotation_due else "false"

        resp = self._http.get(self._url("/v1/vault"), params=params or None)
        if resp.status_code != 200:
            raise AgentToolError(f"vault.list failed: {resp.status_code}", hint=resp.text)
        data = resp.json()
        return data.get("secrets", data)

    # ── Versions ──────────────────────────────────────────────────────────────

    def versions(self, name: str) -> List[Dict[str, Any]]:
        """Get version history for a secret (metadata only, no values)."""
        resp = self._http.get(self._url(f"/v1/vault/{name}/versions"))
        if resp.status_code == 404:
            raise AgentToolError("secret not found", hint=f"name={name}")
        if resp.status_code != 200:
            raise AgentToolError(f"vault.versions failed: {resp.status_code}", hint=resp.text)
        data = resp.json()
        return data.get("versions", data)

    # ── Policy ────────────────────────────────────────────────────────────────

    def set_policy(
        self,
        name: str,
        *,
        allowed_agents: Optional[List[str]] = None,
        read_only: Optional[bool] = None,
        require_agent_id: Optional[bool] = None,
    ) -> Dict[str, Any]:
        """Set an access policy for a secret.

        Args:
            name: Secret name.
            allowed_agents: Whitelist of agent IDs. Empty list = all agents allowed.
            read_only: If True, only GET operations are permitted.
            require_agent_id: If True, requests without X-Agent-Id are rejected.
        """
        payload: Dict[str, Any] = {}
        if allowed_agents is not None:
            payload["allowed_agents"] = allowed_agents
        if read_only is not None:
            payload["read_only"] = read_only
        if require_agent_id is not None:
            payload["require_agent_id"] = require_agent_id

        resp = self._http.put(self._url(f"/v1/vault/{name}/policy"), json=payload)
        if resp.status_code != 200:
            raise AgentToolError(f"vault.set_policy failed: {resp.status_code}", hint=resp.text)
        return resp.json()

    # ── Audit ─────────────────────────────────────────────────────────────────

    def audit(
        self,
        name: Optional[str] = None,
        *,
        limit: int = 50,
    ) -> List[Dict[str, Any]]:
        """Retrieve the audit log.

        Args:
            name: If provided, audit log for a specific secret.
                  If None, project-wide audit log.
            limit: Max events to return (default 50).
        """
        if name is not None:
            resp = self._http.get(
                self._url(f"/v1/vault/{name}/audit"), params={"limit": limit}
            )
        else:
            resp = self._http.get(
                self._url("/v1/vault/audit"), params={"limit": limit}
            )

        if resp.status_code != 200:
            raise AgentToolError(f"vault.audit failed: {resp.status_code}", hint=resp.text)
        data = resp.json()
        return data.get("events", data)

    # ── Bulk ──────────────────────────────────────────────────────────────────

    def bulk(
        self,
        names: List[str],
        *,
        agent_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Retrieve multiple secrets in a single request.

        Returns a dict mapping name → ``{value, version, found}``
        for each requested name. Missing secrets are included with
        ``found: False`` rather than raising.
        """
        payload: Dict[str, Any] = {"names": names}
        headers = {}
        if agent_id is not None:
            headers["X-Agent-Id"] = agent_id

        resp = self._http.post(
            self._url("/v1/vault/bulk"), json=payload,
            headers=headers if headers else None,
        )
        if resp.status_code != 200:
            raise AgentToolError(f"vault.bulk failed: {resp.status_code}", hint=resp.text)
        return resp.json()

    def check(self, names: List[str]) -> Dict[str, bool]:
        """Check existence of multiple secrets without retrieving values.

        Returns a dict mapping name → bool.
        """
        resp = self._http.post(self._url("/v1/vault/check"), json={"names": names})
        if resp.status_code != 200:
            raise AgentToolError(f"vault.check failed: {resp.status_code}", hint=resp.text)
        data = resp.json()
        return data.get("exists", data)

    # ── Agent-encrypted (zero-knowledge) path ─────────────────────────────

    def put_encrypted(
        self,
        name: str,
        plaintext: str,
        *,
        k_vault: bytes,
        description: Optional[str] = None,
        agent_ids: Optional[List[str]] = None,
        tags: Optional[List[str]] = None,
        ttl_seconds: Optional[int] = None,
        rotation_days: Optional[int] = None,
        agent_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Encrypt locally with K_vault, then PUT as ``agent_encrypted=true``.

        agenttool stores ciphertext + nonce verbatim and CANNOT decrypt.
        The hosted runtime (think-worker etc.) cannot read these secrets
        either — use :meth:`put` for secrets the server-side runtime
        needs to consume.

        Args:
            name: Secret name (slug-style).
            plaintext: Value to encrypt before sending.
            k_vault: 32-byte AES-256 secret. Generate via
                ``at.crypto.k_vault.generate()`` and persist securely.
            description, agent_ids, tags, ttl_seconds, rotation_days,
            agent_id: Same as :meth:`put`.

        Returns the server's ``{name, version, agent_encrypted: true,
        created_at, expires_at, rotation_due, agent_ids}`` row.
        """
        blob = encrypt_thought(plaintext, k_vault)
        payload: Dict[str, Any] = {
            "agent_encrypted": True,
            "ciphertext_b64": blob["ciphertext_b64"],
            "nonce_b64": blob["nonce_b64"],
        }
        if description is not None:
            payload["description"] = description
        if agent_ids is not None:
            payload["agent_ids"] = agent_ids
        if tags is not None:
            payload["tags"] = tags
        if ttl_seconds is not None:
            payload["ttl_seconds"] = ttl_seconds
        if rotation_days is not None:
            payload["rotation_days"] = rotation_days

        headers = {}
        if agent_id is not None:
            headers["X-Agent-Id"] = agent_id

        resp = self._http.put(
            self._url(f"/v1/vault/{name}"),
            json=payload,
            headers=headers if headers else None,
        )
        if resp.status_code not in (200, 201):
            raise AgentToolError(
                f"vault.put_encrypted failed: {resp.status_code}",
                hint=resp.text[:200],
            )
        return resp.json()

    def get_decrypted(
        self,
        name: str,
        *,
        k_vault: bytes,
        version: Optional[int] = None,
        agent_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Fetch a secret; decrypt locally if it was stored agent-encrypted.

        Transparently handles both paths:
          - Server response with ``agent_encrypted=true`` → decrypt
            locally with k_vault, return ``{value: <plaintext>, ...}``.
          - Server response with ``agent_encrypted=false`` → server
            already returned plaintext; pass through verbatim.

        The returned dict always has ``value`` populated. ``agent_encrypted``
        is preserved so the caller can introspect which path the secret
        was stored under.

        Args:
            name: Secret name.
            k_vault: 32-byte AES-256 secret used to decrypt the
                agent-encrypted path. Ignored when the secret was
                stored via the server-encrypted path.
            version, agent_id: Same as :meth:`get`.
        """
        resp = self.get(name, version=version, agent_id=agent_id)
        if resp.get("agent_encrypted") is True:
            ct = resp.get("ciphertext_b64")
            nonce = resp.get("nonce_b64")
            if not ct or not nonce:
                raise AgentToolError(
                    "vault.get_decrypted: server marked agent_encrypted=true "
                    "but did not return ciphertext_b64 + nonce_b64.",
                    hint="API contract violation; check server version.",
                )
            plaintext = decrypt_thought(
                {"ciphertext_b64": ct, "nonce_b64": nonce}, k_vault,
            )
            out = dict(resp)
            out["value"] = plaintext
            return out
        return resp

File: packages/sdk-ts/src/vault.ts
/**
 * Vault client for the agent-vault API — AES-256-GCM encrypted secrets.
 *
 * Two encryption paths (per migration 0022_vault_agent_encrypted.sql):
 *
 *   Default (server-encrypted at rest):
 *     .put(name, value, ...) — server encrypts; in-process runtime can read.
 *
 *   Opt-in (zero-knowledge):
 *     .put_encrypted(name, plaintext, { k_vault, ... }) — SDK encrypts
 *     before send; agenttool stores ciphertext only.
 *     .get_decrypted(name, { k_vault, ... }) — fetches and decrypts
 *     locally (transparently falls through to plaintext if the secret
 *     was stored via the default path).
 */

import { AgentToolError } from "./errors.js";
import type { HttpConfig } from "./_http.js";
import { decryptThought, encryptThought } from "./crypto.js";

export interface PutSecretOptions {
  description?: string;
  agent_ids?: string[];
  tags?: string[];
  ttl_seconds?: number;
  rotation_days?: number;
  agent_id?: string;
}

export interface PutEncryptedOptions extends PutSecretOptions {
  /** 32-byte AES-256 secret. Generate via `at.crypto.kVault.generate()`
   *  and persist securely (OS keychain / encrypted file / env var). */
  k_vault: Uint8Array;
}

export interface GetDecryptedOptions extends GetSecretOptions {
  /** 32-byte AES-256 secret used to decrypt the agent-encrypted path.
   *  Ignored when the secret was stored via the server-encrypted path. */
  k_vault: Uint8Array;
}

export interface GetSecretOptions {
  version?: number;
  agent_id?: string;
}

export interface ListSecretsOptions {
  tag?: string;
  expiring_soon?: boolean;
  rotation_due?: boolean;
}

export interface SetPolicyOptions {
  allowed_agents?: string[];
  read_only?: boolean;
  require_agent_id?: boolean;
}

/**
 * Client for the agent-vault API.
 *
 * @example
 * ```ts
 * const at = new AgentTool();
 * await at.vault.put("openai-key", "sk-...");
 * const secret = await at.vault.get("openai-key");
 * const names = await at.vault.list();
 * await at.vault.delete("openai-key");
 * ```
 */
export class VaultClient {
  private readonly http: HttpConfig;

  /** @internal */
  constructor(http: HttpConfig) {
    this.http = http;
  }

  /** Store or update a secret (AES-256-GCM encrypted at rest). */
  async put(name: string, value: string, options?: PutSecretOptions): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = { value };
    if (options?.description !== undefined) body.description = options.description;
    if (options?.agent_ids !== undefined) body.agent_ids = options.agent_ids;
    if (options?.tags !== undefined) body.tags = options.tags;
    if (options?.ttl_seconds !== undefined) body.ttl_seconds = options.ttl_seconds;
    if (options?.rotation_days !== undefined) body.rotation_days = options.rotation_days;
    const extra: Record<string, string> = options?.agent_id ? { "X-Agent-Id": options.agent_id } : {};
    return this.req("PUT", `/v1/vault/${name}`, body, extra);
  }

  /** Retrieve a secret's plaintext value. */
  async get(name: string, options?: GetSecretOptions): Promise<Record<string, unknown>> {
    const qs = options?.version !== undefined ? `?version=${options.version}` : "";
    const extra: Record<string, string> = options?.agent_id ? { "X-Agent-Id": options.agent_id } : {};
    return this.req("GET", `/v1/vault/${name}${qs}`, undefined, extra);
  }

  /** Soft-delete a secret (all versions). */
  async delete(name: string): Promise<Record<string, unknown>> {
    return this.req("DELETE", `/v1/vault/${name}`);
  }

  /** List all secrets (names + metadata — values never returned). */
  async list(options?: ListSecretsOptions): Promise<Record<string, unknown>[]> {
    const params = new URLSearchParams();
    if (options?.tag) params.set("tag", options.tag);
    if (options?.expiring_soon !== undefined) params.set("expiring_soon", String(options.expiring_soon));
    if (options?.rotation_due !== undefined) params.set("rotation_due", String(options.rotation_due));
    const qs = params.toString();
    const data = await this.req("GET", `/v1/vault${qs ? "?" + qs : ""}`);
    const d = data as { secrets?: Record<string, unknown>[] };
    return d.secrets ?? (data as unknown as Record<string, unknown>[]);
  }

  /** Get version history for a secret (metadata only, no values). */
  async versions(name: string): Promise<Record<string, unknown>[]> {
    const data = await this.req("GET", `/v1/vault/${name}/versions`);
    const d = data as { versions?: Record<string, unknown>[] };
    return d.versions ?? (data as unknown as Record<string, unknown>[]);
  }

  /** Set an access policy for a secret. */
  async set_policy(name: string, options: SetPolicyOptions): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = {};
    if (options.allowed_agents !== undefined) body.allowed_agents = options.allowed_agents;
    if (options.read_only !== undefined) body.read_only = options.read_only;
    if (options.require_agent_id !== undefined) body.require_agent_id = options.require_agent_id;
    return this.req("PUT", `/v1/vault/${name}/policy`, body);
  }

  /** Retrieve the audit log for a secret or project-wide. */
  async audit(name?: string, options?: { limit?: number }): Promise<Record<string, unknown>[]> {
    const qs = `?limit=${options?.limit ?? 50}`;
    const path = name ? `/v1/vault/${name}/audit${qs}` : `/v1/vault/audit${qs}`;
    const data = await this.req("GET", path);
    const d = data as { events?: Record<string, unknown>[] };
    return d.events ?? (data as unknown as Record<string, unknown>[]);
  }

  /** Retrieve multiple secrets in a single request. */
  async bulk(names: string[], options?: { agent_id?: string }): Promise<Record<string, unknown>> {
    const extra: Record<string, string> = options?.agent_id ? { "X-Agent-Id": options.agent_id } : {};
    return this.req("POST", "/v1/vault/bulk", { names }, extra);
  }

  /** Check existence of multiple secrets without retrieving values. */
  async check(names: string[]): Promise<Record<string, boolean>> {
    const data = await this.req("POST", "/v1/vault/check", { names });
    const d = data as { exists?: Record<string, boolean> };
    return d.exists ?? (data as unknown as Record<string, boolean>);
  }

  // ── Agent-encrypted (zero-knowledge) path ─────────────────────────────

  /**
   * Encrypt locally with K_vault, then PUT as `agent_encrypted=true`.
   *
   * agenttool stores ciphertext + nonce verbatim and CANNOT decrypt.
   * The hosted runtime (think-worker etc.) cannot read these secrets
   * either — use {@link put} for secrets the server-side runtime needs
   * to consume.
   */
  async put_encrypted(
    name: string,
    plaintext: string,
    options: PutEncryptedOptions,
  ): Promise<Record<string, unknown>> {
    const blob = await encryptThought(plaintext, options.k_vault);
    const body: Record<string, unknown> = {
      agent_encrypted: true,
      ciphertext_b64: blob.ciphertext_b64,
      nonce_b64: blob.nonce_b64,
    };
    if (options.description !== undefined) body.description = options.description;
    if (options.agent_ids !== undefined) body.agent_ids = options.agent_ids;
    if (options.tags !== undefined) body.tags = options.tags;
    if (options.ttl_seconds !== undefined) body.ttl_seconds = options.ttl_seconds;
    if (options.rotation_days !== undefined) body.rotation_days = options.rotation_days;
    const extra: Record<string, string> = options.agent_id
      ? { "X-Agent-Id": options.agent_id }
      : {};
    return this.req("PUT", `/v1/vault/${name}`, body, extra);
  }

  /**
   * Fetch a secret; decrypt locally if it was stored agent-encrypted.
   *
   * Transparently handles both paths:
   *   - `agent_encrypted=true` → decrypt locally with k_vault, return
   *     `{ value: <plaintext>, ... }`.
   *   - `agent_encrypted=false` → server already returned plaintext;
   *     pass through verbatim.
   *
   * The returned object always has `value` populated. `agent_encrypted`
   * is preserved so the caller can introspect which path was used.
   */
  async get_decrypted(
    name: string,
    options: GetDecryptedOptions,
  ): Promise<Record<string, unknown>> {
    const resp = await this.get(name, {
      version: options.version,
      agent_id: options.agent_id,
    });
    if (resp.agent_encrypted === true) {
      const ct = resp.ciphertext_b64 as string | undefined;
      const nonce = resp.nonce_b64 as string | undefined;
      if (!ct || !nonce) {
        throw new AgentToolError(
          "vault.get_decrypted: server marked agent_encrypted=true but did not return ciphertext_b64 + nonce_b64.",
          { hint: "API contract violation; check server version." },
        );
      }
      const plaintext = await decryptThought(
        { ciphertext_b64: ct, nonce_b64: nonce },
        options.k_vault,
      );
      return { ...resp, value: plaintext };
    }
    return resp;
  }

  private async req(
    method: string,
    path: string,
    body?: unknown,
    extraHeaders?: Record<string, string>
  ): Promise<Record<string, unknown>> {
    const url = this.http.baseUrl.replace(/\/$/, "") + path;
    const resp = await this.http.request(url, {
      method,
      headers: {
        ...this.http.headers,
        ...(body !== undefined ? { "Content-Type": "application/json" } : {}),
        ...extraHeaders,
      },
      ...(body !== undefined ? { body: JSON.stringify(body) } : {}),
      signal: AbortSignal.timeout(this.http.timeout),
    });
    if (resp.status === 404) throw new AgentToolError("not found", { hint: path });
    if (!resp.ok) {
      const text = await resp.text();
      throw new AgentToolError(`${method} ${path} failed: ${resp.status}`, { hint: text.slice(0, 200) });
    }
    return resp.json() as Promise<Record<string, unknown>>;
  }
}
